drivers/ipm/ipm_xlnx_ipi.c

```diff
@@ -0,0 +1,262 @@
+	{.ipi_ch_bit = IPI_CH2_BIT},  /* IPI CH ID 2 - Default RPU1 */
+	{.ipi_ch_bit = IPI_CH3_BIT},  /* IPI CH ID 3 - Default PMU0 */
+	{.ipi_ch_bit = IPI_CH4_BIT},  /* IPI CH ID 4 - Default PMU1 */
+	{.ipi_ch_bit = IPI_CH5_BIT},  /* IPI CH ID 5 - Default PMU2 */
+	{.ipi_ch_bit = IPI_CH6_BIT},  /* IPI CH ID 6 - Default PMU3 */
+	{.ipi_ch_bit = IPI_CH7_BIT},  /* IPI CH ID 7 - Default PL0 */
+	{.ipi_ch_bit = IPI_CH8_BIT},  /* IPI CH ID 8 - Default PL1 */
+	{.ipi_ch_bit = IPI_CH9_BIT},  /* IPI CH ID 9 - Default PL2 */
+	{.ipi_ch_bit = IPI_CH10_BIT}, /* IPI CH ID 10 - Default PL3 */
+};
+
+struct xlnx_ipi_config {
+	uint32_t ipi_ch_bit;
+	uint32_t host_ipi_reg;
+	int (*xlnx_ipi_config_func)(const struct device *dev);
+	const struct device **cdev_list;
+	int num_cdev;
+};
+
+struct xlnx_ipi_child_data {
+	bool enabled;
+	ipm_callback_t ipm_callback;
+	void *user_data;
+};
+
+struct xlnx_ipi_child_config {
+	const char *node_id;
+	uint32_t local_request_region;
+	uint32_t local_response_region;
+	uint32_t remote_request_region;
+	uint32_t remote_response_region;
+	uint32_t host_ipi_reg;
+	uint32_t remote_ipi_id;
+	uint32_t remote_ipi_ch_bit;
+};
+
+static void xlnx_mailbox_rx_isr(const struct device *dev)
+{
+	const struct xlnx_ipi_config *config;
+	const struct device **cdev_list;
+	const struct xlnx_ipi_child_config *cdev_conf;
+	const struct xlnx_ipi_child_data *cdev_data;
+	uint8_t ipi_buf[XLNX_IPI_MAX_BUF_SIZE_BYTES + sizeof(struct xlnx_ipi_data)];
+	int num_cdev;
+	struct xlnx_ipi_data *msg;
+	const struct device *cdev;
+	uint32_t remote_ipi_ch_bit;
+	int i, j;
+
+	config = dev->config;
+	cdev_list = config->cdev_list;
+	num_cdev = config->num_cdev;
+	msg = (struct xlnx_ipi_data *)ipi_buf;
+	for (i = 0; i < num_cdev; i++) {
+		cdev = cdev_list[i];
+		cdev_conf = cdev->config;
+		cdev_data = cdev->data;
+
+		if (!cdev_data->enabled) {
+			continue;
+		}
+
+		remote_ipi_ch_bit = cdev_conf->remote_ipi_ch_bit;
+		if (!sys_test_bit(config->host_ipi_reg + IPI_ISR, remote_ipi_ch_bit)) {
+			continue;
+		}
+
+		msg->len = XLNX_IPI_MAX_BUF_SIZE_BYTES;
+		msg->user_data = cdev_data->user_data;
+		for (j = 0; j < XLNX_IPI_MAX_BUF_SIZE_BYTES; j++) {
+			msg->data[j] = sys_read8(cdev_conf->remote_request_region + j);
+		}
+		if (cdev_data->ipm_callback) {
+			cdev_data->ipm_callback(cdev, cdev_data->user_data,
+						cdev_conf->remote_ipi_id, msg);
+		}
+		sys_set_bit(config->host_ipi_reg + IPI_ISR, remote_ipi_ch_bit);
+	}
+}
+
+static int xlnx_ipi_child_init(const struct device *dev)
+{
+	const struct xlnx_ipi_child_config *cdev_conf;
+
+	cdev_conf = dev->config;
+
+	return 0;
+}
+
+static int xlnx_ipi_send(const struct device *ipmdev, int wait, uint32_t id, const void *data,
+			 int size)
+{
+	const uint8_t *msg = (uint8_t *)data;
+	const struct xlnx_ipi_child_config *config = ipmdev->config;
+	unsigned int key;
+	int i, obs_bit;
+
+	ARG_UNUSED(id);
+
+	if (size > XLNX_IPI_MAX_BUF_SIZE_BYTES) {
+		return -EMSGSIZE;
+	}
+
+	key = irq_lock();
```

Comment:
any reason to use `irq_lock` over something SMP safe like using a spinlock?

Reply:
Hi Carlo, Thanks for reviews.

Actually this driver can be used in SMP mode or AMP mode.
For AMP mode, we may not need spinlocks. For SMP mode we will need spinlocks. I will add that support in separate patch.

Thanks.

Reply:
@carlocaione @uLipe 

I tried added spinlocks, but didn't work as expected. ipi_send can be called from interrupt context.  Instead I will have to add workqueue support for each child's callback that are currently being called from ISR. This will be taken care later when SMP use cases are enabled. For now would like to go ahead with simple irq_lock();